File: src/books/metadata-extractor.ts
import { ExternalMetadataService, type ExternalBookMetadata } from "./external-metadata-service.ts";

/**
 * MetadataExtractor - Extracts metadata from various ebook formats
 * Supports EPUB, PDF, and basic file information extraction
 * Enhanced with external API lookups for rich metadata
 */

export interface BookMetadata {
    title?: string;
    author?: string;
    isbn?: string;
    publisher?: string;
    publicationDate?: string;
    description?: string;
    language?: string;
    pageCount?: number;
    fileSize?: number;
    fileFormat?: string;
    coverImagePath?: string;
    coverImageUrl?: string; // From external APIs
    genre?: string[];
    series?: string;
    seriesNumber?: number;
    rating?: number;
    ratingsCount?: number;
    categories?: string[];
    maturityRating?: string;
}

export class MetadataExtractor {
    private externalMetadataService: ExternalMetadataService;
    
    constructor(googleBooksApiKey?: string) {
        this.externalMetadataService = new ExternalMetadataService(googleBooksApiKey);
    }
    
    /**
     * Extract metadata from a file based on its format, enhanced with external API data
     */
    async extractMetadata(filePath: string, enableExternalLookup: boolean = true): Promise<BookMetadata> {
        const metadata: BookMetadata = {};
        
        // Get basic file information
        metadata.fileFormat = this.getFileFormat(filePath);
        metadata.fileSize = await this.getFileSize(filePath);
        
        try {
            switch (metadata.fileFormat?.toLowerCase()) {
                case 'epub':
                    Object.assign(metadata, await this.extractEpubMetadata(filePath));
                    break;
                case 'pdf':
                    Object.assign(metadata, await this.extractPdfMetadata(filePath));
                    break;
                case 'mobi':
                case 'azw3':
                    Object.assign(metadata, await this.extractMobiMetadata(filePath));
                    break;
                default:
                    // For unsupported formats, extract what we can from filename/path
                    Object.assign(metadata, this.extractFromFilename(filePath));
                    break;
            }
        } catch (error) {
            console.warn(`Failed to extract metadata from ${filePath}:`, error);
            // Fallback to filename extraction
            Object.assign(metadata, this.extractFromFilename(filePath));
        }
        
        // Enhance with external API data if enabled
        if (enableExternalLookup) {
            await this.enhanceWithExternalMetadata(metadata, filePath);
        }
        
        return metadata;
    }
    
    /**
     * Enhance metadata with external API data
     */
    private async enhanceWithExternalMetadata(metadata: BookMetadata, filePath: string): Promise<void> {
        try {
            let externalData: ExternalBookMetadata | null = null;
            
            // Try ISBN lookup first if we have an ISBN
            if (metadata.isbn) {
                console.log(`🔍 Looking up ISBN ${metadata.isbn} for enhanced metadata...`);
                externalData = await this.externalMetadataService.lookupByISBN(metadata.isbn);
            }
            
            // If no ISBN result, try title/author search
            if (!externalData && metadata.title && metadata.author) {
                console.log(`🔍 Searching "${metadata.title}" by ${metadata.author} for enhanced metadata...`);
                const searchResults = await this.externalMetadataService.searchByTitle(metadata.title, metadata.author);
                if (searchResults.length > 0) {
                    externalData = searchResults[0]; // Take the best match
                }
            }
            
            // If still no result, try title-only search
            if (!externalData && metadata.title) {
                console.log(`🔍 Searching "${metadata.title}" (title only) for enhanced metadata...`);
                const searchResults = await this.externalMetadataService.searchByTitle(metadata.title);
                if (searchResults.length > 0) {
                    externalData = searchResults[0]; // Take the best match
                }
            }
            
            if (externalData) {
                console.log(`✓ Found enhanced metadata for "${metadata.title || filePath}"`);
                
                // Merge external data with existing metadata (prefer external data for rich fields)
                metadata.title = metadata.title || externalData.title;
                metadata.author = metadata.author || externalData.authors?.[0];
                metadata.isbn = metadata.isbn || externalData.isbn;
                metadata.publisher = externalData.publisher || metadata.publisher;
                metadata.publicationDate = externalData.publishedDate || metadata.publicationDate;
                metadata.description = externalData.description || metadata.description;
                metadata.pageCount = externalData.pageCount || metadata.pageCount;
                metadata.language = metadata.language || externalData.language;
                metadata.coverImageUrl = externalData.coverImageUrl;
                metadata.rating = externalData.rating;
                metadata.ratingsCount = externalData.ratingsCount;
                metadata.categories = externalData.categories;
                metadata.maturityRating = externalData.maturityRating;
                
                // Enhanced genre detection using external categories
                if (externalData.categories && externalData.categories.length > 0) {
                    const detectedGenres = this.externalMetadataService.detectGenresFromCategories(externalData.categories);
                    metadata.genre = [...new Set([...(metadata.genre || []), ...detectedGenres])];
                }
                
                // Enhanced series detection
                if (externalData.series) {
                    metadata.series = externalData.series;
                    metadata.seriesNumber = externalData.seriesNumber;
                }
                
                console.log(`📚 Enhanced metadata: ${metadata.genre?.length || 0} genres, ${metadata.description ? 'description' : 'no description'}, ${metadata.coverImageUrl ? 'cover image' : 'no cover'}`);
            } else {
                console.log(`ℹ️  No external metadata found for "${metadata.title || filePath}"`);
            }
            
        } catch (error) {
            console.warn(`Failed to enhance metadata with external APIs for ${filePath}:`, error);
            // Continue without external enhancement
        }
    }

    /**
     * Extract metadata from EPUB files
     */
    private async extractEpubMetadata(filePath: string): Promise<Partial<BookMetadata>> {
        // For now, we'll implement a basic version
        // In a full implementation, you'd want to use a library like epub-parser
        // or extract the OPF file from the EPUB zip structure
        
        try {
            // This is a placeholder - real implementation would:
            // 1. Read EPUB as ZIP file
            // 2. Parse META-INF/container.xml to find OPF file
            // 3. Parse OPF file for Dublin Core metadata
            // 4. Extract cover image if present
            
            const metadata: Partial<BookMetadata> = {};
            
            // For demonstration, let's extract what we can from the filename
            const filenameMetadata = this.extractFromFilename(filePath);
            Object.assign(metadata, filenameMetadata);
            
            // Add EPUB-specific defaults
            metadata.language = metadata.language || 'en';
            
            return metadata;
        } catch (error) {
            console.error('Error extracting EPUB metadata:', error);
            return this.extractFromFilename(filePath);
        }
    }

    /**
     * Extract metadata from PDF files
     */
    private async extractPdfMetadata(filePath: string): Promise<Partial<BookMetadata>> {
        try {
            // This is a placeholder - real implementation would use a PDF library
            // to read PDF metadata like Title, Author, Subject, Creator, etc.
            
            const metadata: Partial<BookMetadata> = {};
            
            // For demonstration, extract from filename
            const filenameMetadata = this.extractFromFilename(filePath);
            Object.assign(metadata, filenameMetadata);
            
            return metadata;
        } catch (error) {
            console.error('Error extracting PDF metadata:', error);
            return this.extractFromFilename(filePath);
        }
    }

    /**
     * Extract metadata from MOBI/AZW3 files
     */
    private async extractMobiMetadata(filePath: string): Promise<Partial<BookMetadata>> {
        try {
            // This is a placeholder - real implementation would parse MOBI headers
            
            const metadata: Partial<BookMetadata> = {};
            
            // For demonstration, extract from filename
            const filenameMetadata = this.extractFromFilename(filePath);
            Object.assign(metadata, filenameMetadata);
            
            return metadata;
        } catch (error) {
            console.error('Error extracting MOBI metadata:', error);
            return this.extractFromFilename(filePath);
        }
    }

    /**
     * Extract metadata from filename and directory structure
     * Handles both: /books/Author/Book Title/file.ext and /books/Author/file.ext
     */
    private extractFromFilename(filePath: string): Partial<BookMetadata> {
        const metadata: Partial<BookMetadata> = {};
        
        // Extract from path structure
        const pathParts = filePath.split('/');
        const filename = pathParts[pathParts.length - 1];
        const filenameWithoutExt = filename.split('.').slice(0, -1).join('.');
        
        // Find the "books" directory in the path to understand structure
        const booksIndex = pathParts.findIndex(part => part.toLowerCase() === 'books');
        
        if (booksIndex !== -1 && booksIndex < pathParts.length - 2) {
            const authorName = pathParts[booksIndex + 1];
            const nextPart = pathParts[booksIndex + 2];
            
            // Case 1: /books/Author/BookTitle/file.ext (3-level structure)
            if (nextPart !== filename && booksIndex < pathParts.length - 3) {
                metadata.author = this.cleanString(authorName);
                metadata.title = this.cleanString(nextPart);
            }
            // Case 2: /books/Author/file.ext (2-level structure)
            else {
                metadata.author = this.cleanString(authorName);
                metadata.title = this.cleanString(filenameWithoutExt);
            }
        } else {
            // Fallback: just use filename
            metadata.title = this.cleanString(filenameWithoutExt);
        }
        
        // Try to extract ISBN from filename or directory names
        metadata.isbn = this.extractISBN(filePath);
        
        // Try to detect series information
        const seriesInfo = this.extractSeriesInfo(metadata.title || '');
        if (seriesInfo) {
            metadata.series = seriesInfo.series;
            metadata.seriesNumber = seriesInfo.number;
        }
        
        return metadata;
    }

    /**
     * Get file format from extension
     */
    private getFileFormat(filePath: string): string {
        const ext = filePath.split('.').pop()?.toLowerCase();
        return ext || 'unknown';
    }

    /**
     * Get file size in bytes
     */
    private async getFileSize(filePath: string): Promise<number> {
        try {
            const stat = await Deno.stat(filePath);
            return stat.size;
        } catch (error) {
            console.warn(`Could not get file size for ${filePath}:`, error);
            return 0;
        }
    }

    /**
     * Clean up strings extracted from filenames/paths
     */
    private cleanString(str: string): string {
        return str
            .replace(/[_-]/g, ' ')  // Replace underscores and hyphens with spaces
            .replace(/\s+/g, ' ')   // Replace multiple spaces with single space
            .trim();
    }

    /**
     * Extract ISBN from filename or directory names
     */
    private extractISBN(filePath: string): string | undefined {
        // Look for ISBN-10 or ISBN-13 patterns
        const isbnRegex = /(?:ISBN[-\s]*(?:10|13)?[-\s]*[:\s]*)?((?:97[89][-\s]*)?(?:\d[-\s]*){9}[\dX])/gi;
        const match = filePath.match(isbnRegex);
        
        if (match) {
            // Clean up the ISBN (remove spaces, hyphens)
            return match[0].replace(/[^\dX]/gi, '');
        }
        
        return undefined;
    }

    /**
     * Extract series information from title
     * Looks for patterns like "Book Title (Series Name #3)" or "Series Name 3: Book Title"
     */
    private extractSeriesInfo(title: string): { series: string; number: number } | null {
        // Pattern 1: "Book Title (Series Name #3)"
        const pattern1 = /^(.+?)\s*\((.+?)\s*#(\d+)\)$/i;
        const match1 = title.match(pattern1);
        if (match1) {
            return {
                series: match1[2].trim(),
                number: parseInt(match1[3], 10)
            };
        }
        
        // Pattern 2: "Series Name 3: Book Title"
        const pattern2 = /^(.+?)\s+(\d+):\s*(.+)$/i;
        const match2 = title.match(pattern2);
        if (match2) {
            return {
                series: match2[1].trim(),
                number: parseInt(match2[2], 10)
            };
        }
        
        // Pattern 3: "Series Name Book 3"
        const pattern3 = /^(.+?)\s+(?:book|vol|volume)\s+(\d+)$/i;
        const match3 = title.match(pattern3);
        if (match3) {
            return {
                series: match3[1].trim(),
                number: parseInt(match3[2], 10)
            };
        }
        
        return null;
    }

    /**
     * Detect potential genres from metadata, enhanced with external API data
     */
    detectGenres(metadata: BookMetadata, filePath: string): string[] {
        const genres: string[] = [];
        
        // First, use external API genres if available
        if (metadata.genre && metadata.genre.length > 0) {
            genres.push(...metadata.genre);
        }
        
        // Then, use external categories if available
        if (metadata.categories && metadata.categories.length > 0) {
            const detectedGenres = this.externalMetadataService.detectGenresFromCategories(metadata.categories);
            genres.push(...detectedGenres);
        }
        
        // Fallback to keyword-based detection
        const text = `${metadata.title || ''} ${metadata.description || ''} ${filePath}`.toLowerCase();
        
        const genreKeywords = {
            'Cooking': ['cooking', 'recipe', 'kitchen', 'chef', 'culinary', 'food'],
            'Magic': ['magic', 'illusion', 'trick', 'magician', 'prestidigitation'],
            'Romance': ['romance', 'love', 'romantic'],
            'Mystery': ['mystery', 'detective', 'crime', 'murder'],
            'Fantasy': ['fantasy', 'dragon', 'wizard', 'magic', 'quest'],
            'Sci-Fi': ['science fiction', 'sci-fi', 'space', 'alien', 'future'],
            'Horror': ['horror', 'scary', 'ghost', 'vampire', 'zombie'],
            'Biography': ['biography', 'memoir', 'life of', 'autobiography'],
            'History': ['history', 'historical', 'war', 'ancient'],
            'Self-Help': ['self-help', 'improvement', 'success', 'productivity'],
            'Technology': ['technology', 'programming', 'computer', 'software'],
            'Science': ['science', 'physics', 'chemistry', 'biology', 'research']
        };
        
        for (const [genre, keywords] of Object.entries(genreKeywords)) {
            if (keywords.some(keyword => text.includes(keyword))) {
                genres.push(genre);
            }
        }
        
        // Remove duplicates and return
        return [...new Set(genres)];
    }

    /**
     * Detect content rating based on metadata, enhanced with external API data
     */
    detectContentRating(metadata: BookMetadata, filePath: string): string[] {
        // Use external API content rating if available
        if (metadata.maturityRating) {
            const externalRatings = this.externalMetadataService.detectContentRating(metadata);
            if (externalRatings.length > 0) {
                return externalRatings;
            }
        }
        
        // Fallback to keyword-based detection
        const ratings: string[] = [];
        const text = `${metadata.title || ''} ${metadata.description || ''} ${filePath}`.toLowerCase();
        
        const nsfwKeywords = ['adult', 'explicit', 'erotic', 'xxx', 'nsfw'];
        const matureKeywords = ['mature', 'violence', 'blood', 'gore'];
        const teenKeywords = ['young adult', 'teen', 'teenager'];
        const childrenKeywords = ['children', 'kids', 'juvenile'];
        
        if (nsfwKeywords.some(keyword => text.includes(keyword))) {
            ratings.push('NSFW');
        } else if (matureKeywords.some(keyword => text.includes(keyword))) {
            ratings.push('Mature');
        } else if (teenKeywords.some(keyword => text.includes(keyword))) {
            ratings.push('Teen');
        } else if (childrenKeywords.some(keyword => text.includes(keyword))) {
            ratings.push('All Ages');
        } else {
            ratings.push('All Ages');
        }
        
        return ratings;
    }
}

File: src/metadata_extractor.zig
const std = @import("std");
const openlibrary = @import("openlibrary.zig");
const Config = @import("./config.zig").Config;
const build = @import("build.zig.zon");

pub const BookMetadata = struct {
    title: ?[]const u8 = null,
    author: ?[]const u8 = null,
    isbn: ?[]const u8 = null,
    publisher: ?[]const u8 = null,
    publication_date: ?[]const u8 = null,
    description: ?[]const u8 = null,
    language: ?[]const u8 = null,
    page_count: ?u32 = null,
    file_size: u64 = 0,
    file_format: ?[]const u8 = null,
    series: ?[]const u8 = null,
    series_number: ?u32 = null,
    cover_image_url: ?[]const u8 = null,

    pub fn deinit(self: *BookMetadata, allocator: std.mem.Allocator) void {
        if (self.title) |t| allocator.free(t);
        if (self.author) |a| allocator.free(a);
        if (self.isbn) |i| allocator.free(i);
        if (self.publisher) |p| allocator.free(p);
        if (self.publication_date) |pd| allocator.free(pd);
        if (self.description) |d| allocator.free(d);
        if (self.language) |l| allocator.free(l);
        if (self.file_format) |ff| allocator.free(ff);
        if (self.series) |s| allocator.free(s);
        if (self.cover_image_url) |c| allocator.free(c);
    }
};

const SeriesInfo = struct {
    series: []const u8,
    number: u32,
};

pub const MetadataExtractor = struct {
    allocator: std.mem.Allocator,
    ol_client: ?openlibrary.OpenLibraryClient,
    enable_external_lookup: bool,

    pub fn init(allocator: std.mem.Allocator, enable_external_lookup: bool, config: Config) !MetadataExtractor {
        return .{
            .allocator = allocator,
            .ol_client = if (enable_external_lookup) openlibrary.OpenLibraryClient.init(
                allocator,
                config.user_agent,
            ) else null,
            .enable_external_lookup = enable_external_lookup,
        };
    }

    pub fn extractMetadata(self: *MetadataExtractor, file_path: []const u8) !BookMetadata {
        var metadata = BookMetadata{};

        // Get file format from extension
        metadata.file_format = try self.getFileFormat(file_path);

        // Get file size
        metadata.file_size = try self.getFileSize(file_path);

        // Extract from filename/path (works for all formats as fallback)
        try self.extractFromFilename(file_path, &metadata);
        std.log.debug("Extracted metadata from filename: {}", .{metadata});

        // If we have an ISBN and external lookup is enabled, try to enrich with OpenLibrary
        if (self.enable_external_lookup and metadata.isbn != null) {
            if (self.ol_client) |*client| {
                self.enrichWithOpenLibrary(client, &metadata) catch |err| {
                    std.log.warn("Failed to enrich metadata from OpenLibrary: {}", .{err});
                    // Continue with local metadata
                };
            }
        }

        // TODO: In future, implement format-specific extraction:
        // - EPUB: Parse OPF file from ZIP structure
        // - PDF: Extract PDF metadata
        // - MOBI/AZW3: Parse MOBI headers

        return metadata;
    }

    fn enrichWithOpenLibrary(self: *MetadataExtractor, client: *openlibrary.OpenLibraryClient, metadata: *BookMetadata) !void {
        const isbn = metadata.isbn orelse return;

        std.log.info("🔍 Enriching metadata for ISBN: {s}", .{isbn});

        if (try client.lookupByISBN(isbn)) |*external_meta| {
            defer {
                var mut_meta = external_meta.*;
                mut_meta.deinit(self.allocator);
            }

            // Enrich title if external data is better
            if (external_meta.title) |ext_title| {
                if (metadata.title == null or ext_title.len > 0) {
                    if (metadata.title) |old_title| self.allocator.free(old_title);
                    metadata.title = try self.allocator.dupe(u8, ext_title);
                }
            }

            // Enrich author if available
            if (external_meta.authors) |authors| {
                if (authors.len > 0) {
                    if (metadata.author) |old_author| self.allocator.free(old_author);
                    metadata.author = try self.allocator.dupe(u8, authors[0]);
                }
            }

            // Add publisher
            if (external_meta.publisher) |publisher| {
                if (metadata.publisher) |old_pub| self.allocator.free(old_pub);
                metadata.publisher = try self.allocator.dupe(u8, publisher);
            }

            // Add publication date
            if (external_meta.published_date) |date| {
                if (metadata.publication_date) |old_date| self.allocator.free(old_date);
                metadata.publication_date = try self.allocator.dupe(u8, date);
            }

            // Add description
            if (external_meta.description) |desc| {
                if (metadata.description) |old_desc| self.allocator.free(old_desc);
                metadata.description = try self.allocator.dupe(u8, desc);
            }

            // Add page count
            if (external_meta.page_count) |pages| {
                metadata.page_count = pages;
            }

            // Add language
            if (external_meta.language) |lang| {
                if (metadata.language) |old_lang| self.allocator.free(old_lang);
                metadata.language = try self.allocator.dupe(u8, lang);
            }

            // Add cover image URL
            if (external_meta.cover_image_url) |cover| {
                if (metadata.cover_image_url) |old_cover| self.allocator.free(old_cover);
                metadata.cover_image_url = try self.allocator.dupe(u8, cover);
            }

            std.log.info("✨ Enriched metadata: \"{s}\" by {s}", .{
                metadata.title orelse "Unknown",
                metadata.author orelse "Unknown",
            });
        }
    }

    fn getFileFormat(self: *MetadataExtractor, file_path: []const u8) ![]const u8 {
        var i = file_path.len;
        while (i > 0) {
            i -= 1;
            if (file_path[i] == '.') {
                const ext = file_path[i + 1 ..];
                var lower = try self.allocator.alloc(u8, ext.len);
                for (ext, 0..) |c, idx| {
                    lower[idx] = std.ascii.toLower(c);
                }
                return lower;
            }
        }
        return try self.allocator.dupe(u8, "unknown");
    }

    fn getFileSize(self: *MetadataExtractor, file_path: []const u8) !u64 {
        _ = self;
        const file = try std.fs.cwd().openFile(file_path, .{});
        defer file.close();
        const stat = try file.stat();
        return stat.size;
    }

    fn extractFromFilename(self: *MetadataExtractor, file_path: []const u8, metadata: *BookMetadata) !void {
        // Split path into parts
        var path_parts: std.ArrayList([]const u8) = .empty;
        defer path_parts.deinit(self.allocator);

        var it = std.mem.tokenizeScalar(u8, file_path, '/');
        while (it.next()) |part| {
            try path_parts.append(self.allocator, part);
        }

        if (path_parts.items.len == 0) return;

        // Get filename without extension
        const filename = path_parts.items[path_parts.items.len - 1];
        const filename_no_ext = self.getFilenameWithoutExt(filename);

        // Find "books" directory in path
        var books_index: ?usize = null;
        for (path_parts.items, 0..) |part, i| {
            if (std.ascii.eqlIgnoreCase(part, "books")) {
                books_index = i;
                break;
            }
        }

        if (books_index) |idx| {
            // Case 1: /books/Author/BookTitle/file.ext (3-level)
            if (idx + 2 < path_parts.items.len - 1) {
                const author_name = path_parts.items[idx + 1];
                const book_title = path_parts.items[idx + 2];
                metadata.author = try self.cleanString(author_name);
                metadata.title = try self.cleanString(book_title);
            }
            // Case 2: /books/Author/file.ext (2-level)
            else if (idx + 1 < path_parts.items.len) {
                const author_name = path_parts.items[idx + 1];
                metadata.author = try self.cleanString(author_name);
                metadata.title = try self.cleanString(filename_no_ext);
            }
        } else {
            // Fallback: use filename as title
            metadata.title = try self.cleanString(filename_no_ext);
        }

        // Try to extract ISBN
        metadata.isbn = self.extractISBN(file_path);

        // Try to extract series info
        if (metadata.title) |title| {
            if (self.extractSeriesInfo(title)) |series_info| {
                metadata.series = try self.allocator.dupe(u8, series_info.series);
                metadata.series_number = series_info.number;
            }
        }
    }

    fn getFilenameWithoutExt(self: *MetadataExtractor, filename: []const u8) []const u8 {
        _ = self;
        var i = filename.len;
        while (i > 0) {
            i -= 1;
            if (filename[i] == '.') {
                return filename[0..i];
            }
        }
        return filename;
    }

    fn cleanString(self: *MetadataExtractor, str: []const u8) ![]const u8 {
        var result = try self.allocator.alloc(u8, str.len);
        var out_idx: usize = 0;
        var last_was_space = false;

        for (str) |c| {
            if (c == '_' or c == '-') {
                if (!last_was_space) {
                    result[out_idx] = ' ';
                    out_idx += 1;
                    last_was_space = true;
                }
            } else if (c == ' ') {
                if (!last_was_space) {
                    result[out_idx] = c;
                    out_idx += 1;
                    last_was_space = true;
                }
            } else {
                result[out_idx] = c;
                out_idx += 1;
                last_was_space = false;
            }
        }

        // Trim trailing spaces
        while (out_idx > 0 and result[out_idx - 1] == ' ') {
            out_idx -= 1;
        }

        // Trim leading spaces
        var start_idx: usize = 0;
        while (start_idx < out_idx and result[start_idx] == ' ') {
            start_idx += 1;
        }

        const cleaned = try self.allocator.dupe(u8, result[start_idx..out_idx]);
        self.allocator.free(result);
        return cleaned;
    }

    fn extractISBN(self: *MetadataExtractor, file_path: []const u8) ?[]const u8 {
        // Extract ISBN-10 or ISBN-13 from filename
        // Supports hyphens, underscores, spaces as separators
        // Supports ISBN-10 with trailing 'X' check digit
        const basename = std.fs.path.basename(file_path);

        // Remove extension
        const name_without_ext = if (std.mem.lastIndexOf(u8, basename, ".")) |idx|
            basename[0..idx]
        else
            basename;

        var digit_buffer: [20]u8 = undefined;
        var digit_count: usize = 0;

        for (name_without_ext) |c| {
            if (std.ascii.isDigit(c)) {
                if (digit_count < digit_buffer.len) {
                    digit_buffer[digit_count] = c;
                    digit_count += 1;
                }
                continue;
            }

            // Support ISBN-10 with trailing X (check digit)
            if ((c == 'x' or c == 'X') and digit_count == 9) {
                if (digit_count < digit_buffer.len) {
                    digit_buffer[digit_count] = 'X';
                    digit_count += 1;
                }
                continue;
            }

            // Skip separators (hyphens, underscores, spaces, dots)
            if (c == '-' or c == '_' or c == ' ' or c == '.') {
                continue;
            }

            // Non-separator, non-digit character - check if we have a complete ISBN
            if (digit_count == 10 or digit_count == 13) {
                const isbn = self.allocator.dupe(u8, digit_buffer[0..digit_count]) catch return null;
                std.log.debug("Extracted ISBN from filename: {s}", .{isbn});
                return isbn;
            }
            digit_count = 0;
        }

        // Check final sequence
        if (digit_count == 10 or digit_count == 13) {
            const isbn = self.allocator.dupe(u8, digit_buffer[0..digit_count]) catch return null;
            std.log.debug("Extracted ISBN from filename: {s}", .{isbn});
            return isbn;
        }

        return null;
    }

    fn extractSeriesInfo(self: *MetadataExtractor, title: []const u8) ?SeriesInfo {
        _ = self;
        // Pattern 1: "Book Title (Series Name #3)"
        // Pattern 2: "Series Name 3: Book Title"
        // Pattern 3: "Series Name Book 3"
        // TODO: Implement series info extraction
        _ = title;
        return null;
    }

    pub fn detectGenres(self: *MetadataExtractor, metadata: *const BookMetadata, file_path: []const u8) ![][]const u8 {
        var genres: std.ArrayList([]const u8) = .empty;

        const text_parts = [_]?[]const u8{
            metadata.title,
            metadata.description,
            file_path,
        };

        var combined: std.ArrayList(u8) = .empty;
        defer combined.deinit(self.allocator);

        for (text_parts) |part| {
            if (part) |p| {
                try combined.appendSlice(self.allocator, p);
                try combined.append(self.allocator, ' ');
            }
        }

        // Convert to lowercase for matching
        for (combined.items) |*c| {
            c.* = std.ascii.toLower(c.*);
        }

        const text = combined.items;

        // Genre keywords
        const GenreKeywords = struct {
            genre: []const u8,
            keywords: []const []const u8,
        };

        const genre_keywords = [_]GenreKeywords{
            .{ .genre = "Cooking", .keywords = &[_][]const u8{ "cooking", "recipe", "kitchen", "chef", "culinary", "food" } },
            .{ .genre = "Magic", .keywords = &[_][]const u8{ "magic", "illusion", "trick", "magician" } },
            .{ .genre = "Romance", .keywords = &[_][]const u8{ "romance", "love", "romantic" } },
            .{ .genre = "Mystery", .keywords = &[_][]const u8{ "mystery", "detective", "crime", "murder" } },
            .{ .genre = "Fantasy", .keywords = &[_][]const u8{ "fantasy", "dragon", "wizard", "quest" } },
            .{ .genre = "Sci-Fi", .keywords = &[_][]const u8{ "science fiction", "sci-fi", "space", "alien", "future" } },
            .{ .genre = "Horror", .keywords = &[_][]const u8{ "horror", "scary", "ghost", "vampire", "zombie" } },
            .{ .genre = "Biography", .keywords = &[_][]const u8{ "biography", "memoir", "life of", "autobiography" } },
            .{ .genre = "History", .keywords = &[_][]const u8{ "history", "historical", "war", "ancient" } },
            .{ .genre = "Self-Help", .keywords = &[_][]const u8{ "self-help", "improvement", "success", "productivity" } },
            .{ .genre = "Technology", .keywords = &[_][]const u8{ "technology", "programming", "computer", "software" } },
            .{ .genre = "Science", .keywords = &[_][]const u8{ "science", "physics", "chemistry", "biology", "research" } },
        };

        for (genre_keywords) |gk| {
            for (gk.keywords) |keyword| {
                if (std.mem.indexOf(u8, text, keyword)) |_| {
                    try genres.append(self.allocator, try self.allocator.dupe(u8, gk.genre));
                    break;
                }
            }
        }

        return genres.toOwnedSlice(self.allocator);
    }
};

// Tests
const testing = std.testing;

test "extractISBN extracts 13-digit ISBN with hyphens" {
    var extractor = try MetadataExtractor.init(
        testing.allocator,
        false,
        Config.init(),
    );
    const sample = "/path/to/978-1-098-16220-7.epub";
    const isbn = extractor.extractISBN(sample);
    try testing.expect(isbn != null);
    defer if (isbn) |i| testing.allocator.free(i);
    try testing.expectEqualStrings("9781098162207", isbn.?);
}

test "extractISBN supports ISBN-10 with X suffix" {
    var extractor = try MetadataExtractor.init(
        testing.allocator,
        false,
        Config.init(),
    );
    const sample = "/library/TheBook_123456789X.epub";
    const isbn = extractor.extractISBN(sample);
    try testing.expect(isbn != null);
    defer if (isbn) |i| testing.allocator.free(i);
    try testing.expectEqualStrings("123456789X", isbn.?);
}

test "extractISBN returns null when no ISBN present" {
    var extractor = try MetadataExtractor.init(
        testing.allocator,
        false,
        Config.init(),
    );
    const sample = "/books/no-isbn-book.pdf";
    const isbn = extractor.extractISBN(sample);
    try testing.expect(isbn == null);
}

test "extractISBN extracts plain 13-digit ISBN" {
    var extractor = try MetadataExtractor.init(
        testing.allocator,
        false,
        Config.init(),
    );
    const sample = "9781098162207.pdf";
    const isbn = extractor.extractISBN(sample);
    try testing.expect(isbn != null);
    defer if (isbn) |i| testing.allocator.free(i);
    try testing.expectEqualStrings("9781098162207", isbn.?);
}

test "extractISBN extracts ISBN with underscores and spaces" {
    var extractor = try MetadataExtractor.init(
        testing.allocator,
        false,
        Config.init(),
    );
    const sample = "978_1_098_16220_7 - Book Title.pdf";
    const isbn = extractor.extractISBN(sample);
    try testing.expect(isbn != null);
    defer if (isbn) |i| testing.allocator.free(i);
    try testing.expectEqualStrings("9781098162207", isbn.?);
}

test "extractMetadata preserves ISBN through full extraction" {
    var extractor = try MetadataExtractor.init(
        testing.allocator,
        false,
        Config.init(),
    );

    // Test just the filename parsing portion that includes ISBN extraction
    const sample_path = "/books/Author Name/Book Title/978-0-123-45678-9.epub";

    // Since extractMetadata tries to open the file, we'll test the ISBN extraction directly
    const isbn = extractor.extractISBN(sample_path);
    defer if (isbn) |i| testing.allocator.free(i);

    try testing.expect(isbn != null);
    try testing.expectEqualStrings("9780123456789", isbn.?);
}
